File: framework-docs/src/main/java/org/springframework/docs/web/webmvc/mvccontroller/mvcannvalidation/HandlerMethodValidationExceptionVisitor.java
package org.springframework.docs.web.webmvc.mvccontroller.mvcannvalidation;

import org.jspecify.annotations.Nullable;
import org.springframework.validation.method.ParameterErrors;
import org.springframework.validation.method.ParameterValidationResult;
import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.MatrixVariable;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RequestPart;
import org.springframework.web.method.annotation.HandlerMethodValidationException;

public class HandlerMethodValidationExceptionVisitor {

	static void main() {
		// tag::snippet[]
		HandlerMethodValidationException ex = /**/ new HandlerMethodValidationException(null);

		ex.visitResults(new HandlerMethodValidationException.Visitor() {

			@Override
			public void requestHeader(RequestHeader requestHeader, ParameterValidationResult result) {
				// ...
			}

			@Override
			public void requestParam(@Nullable RequestParam requestParam, ParameterValidationResult result) {
				// ...
			}

			@Override
			public void modelAttribute(@Nullable ModelAttribute modelAttribute, ParameterErrors errors) {
				// ...
			}

			// @fold:on // ...
			@Override
			public void requestPart(RequestPart requestPart, ParameterErrors errors) {

			}

			@Override
			public void cookieValue(CookieValue cookieValue, ParameterValidationResult result) {

			}

			@Override
			public void matrixVariable(MatrixVariable matrixVariable, ParameterValidationResult result) {

			}

			@Override
			public void pathVariable(PathVariable pathVariable, ParameterValidationResult result) {

			}

			@Override
			public void requestBody(RequestBody requestBody, ParameterErrors errors) {

			}
			// @fold:off

			@Override
			public void other(ParameterValidationResult result) {
				// ...
			}
		});
		// end::snippet[]
	}

}

File: framework-docs/src/main/kotlin/org/springframework/docs/web/webmvc/mvccontroller/mvcannvalidation/HandlerMethodValidationExceptionVisitor.kt
package org.springframework.docs.web.webmvc.mvccontroller.mvcannvalidation

import org.springframework.context.MessageSourceResolvable
import org.springframework.validation.method.MethodValidationResult
import org.springframework.validation.method.ParameterErrors
import org.springframework.validation.method.ParameterValidationResult
import org.springframework.web.bind.annotation.*
import org.springframework.web.method.annotation.HandlerMethodValidationException
import java.lang.reflect.Method

class HandlerMethodValidationExceptionVisitor {

	fun main() {
		// tag::snippet[]
		val ex: HandlerMethodValidationException =  /**/ HandlerMethodValidationException(EmptyMethodValidationResult())

		ex.visitResults(object : HandlerMethodValidationException.Visitor {

			override fun requestHeader(requestHeader: RequestHeader, result: ParameterValidationResult) {
				// ...
			}

			override fun requestParam(requestParam: RequestParam?, result: ParameterValidationResult) {
				// ...
			}

			override fun modelAttribute(modelAttribute: ModelAttribute?, errors: ParameterErrors) {
				// ...
			}

			// @fold:on // ...
			override fun requestPart(requestPart: RequestPart, errors: ParameterErrors) {
			}

			override fun cookieValue(cookieValue: CookieValue, result: ParameterValidationResult) {
			}

			override fun matrixVariable(matrixVariable: MatrixVariable, result: ParameterValidationResult) {
			}

			override fun pathVariable(pathVariable: PathVariable, result: ParameterValidationResult) {
			}

			override fun requestBody(requestBody: RequestBody, errors: ParameterErrors) {
			}
			// @fold:off

			override fun other(result: ParameterValidationResult) {
				// ...
			}
		})
		// end::snippet[]
	}

	internal class EmptyMethodValidationResult : MethodValidationResult {
		override fun getTarget(): Any {
			TODO()
		}

		override fun getMethod(): Method {
			TODO()
		}

		override fun isForReturnValue(): Boolean {
			TODO()
		}

		override fun getParameterValidationResults(): List<ParameterValidationResult> {
			TODO()
		}

		override fun getCrossParameterValidationResults(): List<MessageSourceResolvable> {
			TODO()
		}

		override fun toString(): String {
			TODO()
		}
	}
}